Step4 리뷰 요청합니다 :)

kwanguuuu opened the pull request:
리뷰어님 안녕하세요.! 조금 텀이 있었지만, step4 리뷰 요청 드려요~
tdd 하는 것 처럼 
- 선 테스트 작성 -> 후 로직 구현 & 검증
행동하기가 쉽지 않네요..

쉽지 않은 이유중 하나가.. 커밋단위 컴파일 에러가 많이 발생하는것도 이유인것 같아요.!
커밋 단위 실행하면, 실행 안되는 케이스가 있을까 걱정이 되는데, 이점 참고 부탁드려요.! 

### 주의
klintFormat 실행이 안됐는데, 마지막 커밋 c10debf193b1eb23e528dace73153567ccca6240
에 클래스명이 잘못됐단걸 알아서.. klintFormat이 실행이 안될것 같습니다  😇

이번에도 좋은 리뷰 부탁드릴게요!

kwanguuuu committed:
[step3-pr] fix:  step3 pr 반영 - 완료된 항목 docs checkbox 적용, 기능요구 변경된 docs 수정

kwanguuuu committed:
[step3-pr] refactor: step3 pr 반영 - InputView 함수명 변경

- pr : https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r871331814

kwanguuuu committed:
[step3-pr] refactor: step3 패키지 삭제, 다른 패키지랑 겹치는 패스가 있어서 이전 구현체들 위치 변경(step1)

- pr : https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r871377055

kwanguuuu committed:
[step3-pr] refactor: RacingGame 생성 방식 수정

- pr : https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r871384681

kwanguuuu committed:
[step3-pr] fix: 테스트함수명 오타 수정

kwanguuuu committed:
[step4] docs: step4 구현기능 관련 docs 작성

kwanguuuu committed:
[step4] feat:  driver 객체 생성에 name 추가, 컴파일 에러 방지 목적으로 dummy string 작성

kwanguuuu committed:
[step4] feat: Name 생성 전략 클래스, 익셉션 클래스 작성. 테스트 작성

kwanguuuu committed:
[step4] feat: Car 객체 생성 방식 변경, carname 추가 주행결과를 int -> moveResult 객체 리턴하도록 수정

kwanguuuu committed:
[step4] feat: Driver 객체 변경 ( Driver -> Drivers ), 객체 생성시 이름 전략 의존하도록 수정

kwanguuuu committed:
[step4] feat: RacingGame 컴파일 에러 수정

kwanguuuu committed:
[step4] feat: InputView, ResultView 수정

kwanguuuu committed:
[step4] feat: MoveResult에 최대이동한 자동차명 얻기 구현

kwanguuuu committed:
[step4] feat: RacingGame 결과 view 구현.

kwanguuuu committed:
[step4] refactor: 패키지 변경, 관련 코드 수정

kwanguuuu committed:
[step4] fix: docs 수정, MoveResults 파일명 수정 ( klintFormat 밸리데이션 )

wisemuji commented on src/test/kotlin/racingcar/car/DriversTest.kt:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    @Test
+    fun `드라이버의 전략이 true면 차가 움직일 수 있다`() {
+        // given
+        val alwayMoveDriver = Drivers(listOf("car"), AlwaysTrueMoveStrategy())
+
+        // when
+        val moveCount = 100
+        repeat(moveCount) {
+            alwayMoveDriver.driveAll()
+        }
+
+        // then
+        assertThat(alwayMoveDriver.getMoveResult()).isEqualTo(100)
+    }
+
+    @Test
+    fun `드라이버의 전략이 false면 차가 움직일 수 없다`() {
```
<img width="563" alt="image" src="https://user-images.githubusercontent.com/32327475/168454226-dee1b6fd-d875-4de6-ac3b-4499ccb5151b.png">
두 테스트가 실패하네요! 확인 부탁드려요!

kwanguuuu replied:
커밋 전에 전체 빌드를 하지 않아서, 테스트 통과실패 케이스가 있는지 확인하지 못한 상태 였네요.
테스트 실패가
기존 반환타입이 Int(이동거리) 에서 MoveResults(레이스 결과 객체) 로 변경되면서 실패하게 된 케이스인데요.

리팩토링 할 때,  최종적으로 만든 클래스에서, `차가 움직일 수 있다`란 값을 드라이버로부터 도출할 수 있는지를 고민하게 되네요..

kwanguuuu replied:
드라이버가 차를 이동명령을 한 후에, 차가 이동했는지를 아는것도 중요한것 같아서, 테스트 코드가 맘에들진 않지만.. 정상 검증 되도록 수정했습니다.
약간 단위 테스트라기보단, Driversm MoveStrategy, Car 세 객채를 함꼐 테스트 하는 것 같아서 기능테스트가 된것 같아요.!
리뷰어님 의견은 어떤지 궁금하네요 ㅎㅎ
https://github.com/next-step/kotlin-racingcar/pull/539/commits/fe49f641e2ee3390001a3cb241e3b994cd89e9c4

wisemuji replied:
MoveStrategy는 고정된 상태(AlwaysTrueMoveStrategy)이므로 제외한다고 하면 Driver와 Car객체를 같이 테스트한다는 생각을 하실 수 있을 것 같은데요, 한 객체 내부에 다른 객체의 인스턴스가 존재하는 경우를 테스트하게 되면 이런 고민이 생길 수 있을 것 같아요. 물론 모든 객체를 완벽히 독립적으로 테스트하는 다른 방법이 있을 수 있겠지만 저는 지금도 충분하다고 생각합니다.

그래도 이 테스트가 통합 테스트의 성격이라고 생각되신다면 지금은 한 테스트에서 driveAll()과 getMoveResults()를 같이 테스트하고 있는데 이걸 분리해보는 시도를 할 수 있을 것 같아요.

wisemuji commented on README.md:
```diff
@@ -15,17 +15,16 @@
 ## step3 기능 요구사항 정리
```
요구사항 계속 유지보수 하시는 모습 👍  좋습니다!

wisemuji commented on src/main/kotlin/racingcar/car/RacingGame.kt:
```diff
@@ -0,0 +1,42 @@
+        return drivers.getMoveResult()
+    }
+
+    fun isPlayable(): Boolean {
+        if (trialCount <= GAME_END_COUNT) {
+            throw GameOverException()
+        }
+        return true
+    }
+
```
https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r871388083 이 코멘트에 대해 고민해보면 좋겠어요!

kwanguuuu replied:
말씀해주신 대로 isPlayable()을 Boolean만 반환하도록 동작을 수정했어요.
https://github.com/next-step/kotlin-racingcar/pull/539/commits/8edb77a8314a2d233980434a7c946bbdd11fde20

수정하면서 느낀게 하나 있는데,
Booleam, exception 두가지를 반환하니까 테스트코드에도 두 경우를 한 테스트 코드에서 검증하려고 하고 있더라구요..
그래서 그 테스트도 분리했습니다.

테스트 코드를 통해서 개선할 부분이 있는 코드구나 라고 느낀것 같아요.. 감사합니다

wisemuji replied:
리뷰를 아주 잘 반영해주셨습니다 😄 💯

wisemuji commented on src/test/kotlin/racingcar/NumberInputTest.kt:
```diff
@@ -1,11 +1,12 @@
 import org.junit.jupiter.params.provider.ValueSource
+import racingcar.utils.InputParser
 
 class NumberInputTest {
```
https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r871399154 이 코멘트도 반영 부탁드립니다!

kwanguuuu replied:
신규 테스트엔 작성하고 기존건 변경하지 않아서, 수정했습니다.
https://github.com/next-step/kotlin-racingcar/pull/539/commits/97a8bba7c987ad7805a81f054260cfb2acbdb06d
p.s
패키지를 보니 assertThrows는 junit에 있고, assertThatThrownBy는 assertj 패키지에 있네요.

junit패키지에 있는 클래스가 코틀린에서 래핑해서 사용하고 있는 클래스다보니, 사용을 권장하시는 걸까요!? 
아니면 junit5에 추가된 기능이라서 권장하시는 걸까요!? (...둘다인가 👀)

https://velog.io/@ghks285/junit5-assertThatThrownBy
관련해서 짧게 의견을 작성해둔 블로그가 있어서 참조는 했는데.. 뭔가 확 와닿지는 않네요
+assertThat을 가장 많이 쓴느데 이건 래핑 클래스가 없나봐요.. assertAll뿐이네요.. :(

wisemuji replied:
사용을 권장드리는 가장 큰 이유는 Reflection 때문인데요, AssertJ 패키지에 있는 assertThatThrownBy는 java API다 보니 kotlin의 IllegalArgumentException을 isInstanceOf로 넘기기 위해선 Reflection을 이용해야 합니다. (`IllegalArgumentException::class.java`)

참고로 isInstanceOf는 이렇게 생겼어요!
```java
  @Override
  public SELF isInstanceOf(Class<?> type) {
    objects.assertIsInstanceOf(info, actual, type);
    return myself;
  }
```

assertThrows의 경우 코틀린으로 작성되어 있기 때문에 Reflection을 사용하지 않아도 됩니다.
참고로 assertThrows는 이렇게 생겼어요!
```kotlin
inline fun <reified T : Throwable> assertThrows(executable: () -> Unit): T {
    val throwable: Throwable? = try {
        executable()
    } catch (caught: Throwable) {
        caught
    } as? Throwable

    return Assertions.assertThrows(T::class.java) {
        if (throwable != null) {
            throw throwable
        }
    }
}
```

wisemuji replied:
Java API 대신 최대한 대체할 수 있는 Kotlin API를 사용해보세요!

kwanguuuu replied:
친절한 설명 감사합니다! :)

wisemuji commented on src/test/kotlin/racingcar/car/NumberGenerateTest.kt:
```diff
@@ -1,8 +1,8 @@
-import step3.racingcar.RandomNumberGenerator
+import racingcar.utils.RandomNumberGenerator
 
 class NumberGenerateTest {
```
https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r871408947 이 코멘트에 대해 고민해보시면 좋을 것 같아요!

kwanguuuu replied:
오 넵 ㅎㅎ랜덤 로직 자체를 테스트 하는건 올바르지 않은것 같아요
공유해주신 sof에 베스트 앤서가, `무엇을 증명하려고 하는지?` 에 대해서 스스로 물어보라고 얘기하네요.
( 그 외에 예시 내용은 번역기..를 통해서 읽엇는데 딱히 와닿진 않았어요!)

---
제가 검증하려고 했던 부분 중 항상 테스트 가능한건
정수 생성 범위를 0-9사이 정수가 맞는지 만 인거 같아요
다시 보니 이게 단위테스트라기 보다는
IntRange로 생성되는 값이 내가 생각한 값이 맞는지에 대한 학습테스트가 맞는 표현인것 같아요

음 그러면 테스트를 다른 방식으로 작성해야 할것 같아요.. 단위테스트가 아닌 intRange 학습테스트를요..
근데 그러면 현재 테스트클래스는 필요가 없겠네요 ! (+ 지금 시점에 굳이 학습테스트 작성 안해도 될것 같아요 !ㅋㅋ)

결론은 이 테스트 클래스는 지우는게 어떨까요

리뷰어님 의견 부탁합니다 🙏🏻

wisemuji replied:
저도 해당 테스트 클래스는 지우는 것을 추천드립니다. ㅎㅎ

조금 더 첨언하자면 랜덤 클래스를 테스트하게 되면 실행 시점에 따라 실행 결과가 동적으로 바뀔 가능성이 있어요. 
항상 테스트가 성공해서 올바른 로직인 줄 알았는데 어쩔 때는 실패하는 케이스가 생길 수 있겠죠.

따라서 비즈니스로직에 랜덤이 사용된다면 최대한 랜덤 로직을 제외한 부분을 테스트하고(광유님께서 하신것처점 전략 패턴을 이용하는게 좋은 방법이 될 수 있겠죠) 랜덤 그 자체를 테스트하는건 지양하는게 좋습니다 🙂

wisemuji commented on src/main/kotlin/racingcar/Main.kt:
```diff
@@ -0,0 +1,22 @@
+
+    val racingGame = RacingGame(carNames, trialCount, moveStrategy, nameStrategy)
+
+    repeat(racingGame.trialCount) {
+        racingGame.play()
+        ResultView.showRecordsAsDash(racingGame.getPlayResults())
+    }
+    ResultView.showRaceWinner(racingGame.getWinners())
```
ResultView의 함수가 매 턴마다 실행되고 있어요.
ResultView가 서버, 호출하는 곳이 클라이언트라고 가정하고 호출 비용을 최소화해보면 어떨까요? 어떤 방법을 시도해볼 수 있을까요?

kwanguuuu replied:
엇 moveResult가 play의 마지막 결과가 아니라, 모든 기록을 갖고 있으면 될거 같은데.. 고민을 좀 해보겠습니다 (PR 요청 범위에서 놓쳤네요. 넘 늦어서 담에 할게요!)

wisemuji replied:
네! 이 부분은 5단계에서 고민해주셔도 좋습니다 👍

wisemuji commented on src/main/kotlin/racingcar/Main.kt:
```diff
@@ -0,0 +1,22 @@
+fun main() {
+    val carNames = InputView.getNumberOfCarByConsole()
+    val trialCount = InputView.getTrialCountByConsole()
+    val moveStrategy = RandomNumberStrategy()
+    val nameStrategy = NameLengthLimitStrategy()
+
+    val racingGame = RacingGame(carNames, trialCount, moveStrategy, nameStrategy)
```
default parameter의 경우 호출을 생략할 수 있습니다! nameStrategy은 지금도 생략이 가능해보입니다

kwanguuuu replied:
ㅎㅎ.. 디게 웃긴게 제가 Drivers객체에만  default parameter만 줬네요 RacingGame엔 주지 않았어요!
이유는.. 
- DriverTest 클래스에서 Driver 객체 생성시 인터페이스 직접구현해서 nameStrategy를 주기가 귀찮았던것 같아요,,
- 습관은 아닌데, nameStrategy는 moveStrategy와 달리 1회성 참조만 하고 프로퍼티로 사용하지 않아서, default값을 주고싶었나봐요 참조하지 않아도 되도록...

이건 조금더 고민하고,., 레이싱게임에 default를 주던가 아니만 삭제하던가 선택하겠습니다...

wisemuji commented on src/main/kotlin/racingcar/car/MoveResults.kt:
```diff
@@ -0,0 +1,18 @@
+package racingcar.car
+
+class MoveResults(
```
5단계를 진행하시게 되면 domain, view, controller의 역할을 각각 나누는 작업을 진행하시게 될 텐데요, MoveResults와 같은 클래스가 과연 이 셋 중에 어떤 곳에 속할지 미리 고민해보셔도 좋을 것 같아요! (여기선 반영하지 않으셔도 됩니다)

kwanguuuu replied:
음 제 생각엔 domain - view 사이의 중간 dto객체로 사용할것 같아요.
driver나 car를 도메인일텐데, view에 직접 노출하는게 편하긴 한데, 좋은방법인지는 의문이 있어서요.

단순 dto또는 VO 라고만 생각하면 결과를 만드는 걸 도메인에서 해주거나 아님 만들어잔 값을 MoveResults가 갖고 있어야 할것 같습니다.
(= moveResult 내에 farthestDistanceCarNames 프로퍼티가 이동이 필요한것 같기도 ⁉️)

리뷰님 의견 궁금하네요!

wisemuji replied:
> driver나 car를 도메인일텐데, view에 직접 노출하는게 편하긴 한데, 좋은방법인지는 의문이 있어서요.

요구사항에 따르면 domain을 view에 직접 노출하도록 해도 큰 문제는 없습니다! 다만 고민이 필요한 부분은 domain이 view를 꼭 알고 있지 않더라도 view의 요구사항에 맞춤화된 데이터를 들고 있는 것은 지양해야 합니다.

> moveResult 내에 farthestDistanceCarNames 프로퍼티가 이동이 필요한것 같기도 ⁉️

저도 같은 의견입니다. MoveResult는 DTO의 역할을 한다고 하더라도 MoveResults의 farthestDistanceCarNames은 우승자를 판별하는 도메인 로직이라는 생각이 들어요. 도메인 로직과 뷰 로직을 분리해보세요!

view -> client, domain -> server라고 생각하시면 더 쉬우실거예요 😉

wisemuji commented on src/main/kotlin/racingcar/car/RacingGame.kt:
```diff
@@ -0,0 +1,42 @@
+import racingcar.car.strategy.MoveStrategy
+import racingcar.car.strategy.NameStrategy
+
+class RacingGame(carNames: List<String>, trialCount: Int, moveStrategy: MoveStrategy, nameStrategy: NameStrategy) {
```
> 단순 이동시키니..
란 엄청 긴 생성자 이름이 나왔네용..
이게 더 맘에 안들어서.. 주생성자로 변경했습니닷!

가독성의 이유로 [secondary constructor](https://kotlinlang.org/docs/classes.html#secondary-constructors)를 사용하지 않으셨군요! 개행을 하는 방법도 있습니다 🙂  secondary constructor를 사용하면 테스트할 때 의도한 데이터만을 가지고 테스트함을 확실히 보장할 수 있다는 장점이 있습니다. 고민해보시고 반영 여부는 선택해주세요!

kwanguuuu replied:
넵.. 이건 약간 취향을 탄것 같습니닷. 개행을 했는데도 파라미터가 여럿이라 흩어져 있는 느낌이더라구요.
파라미터가 여러개 일 때, 파라미터 1개당 1줄로 작성하는게 올바른 방법이죠..!? (그렇게 알고있는데..)

wisemuji replied:
> 파라미터가 여러개 일 때, 파라미터 1개당 1줄로 작성하는게 올바른 방법이죠..!? (그렇게 알고있는데..)

네 저도 한 줄로 작성하기에 너무 길어지는 경우 개행을 권장하는 것으로 알고 있습니다. ([참고](https://kotlinlang.org/docs/coding-conventions.html#class-headers))

https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r872198694 에서 작성해주신 스니펫 기준으로 이렇게 개행할 수 있겠네요!
```kotlin
 constructor(
    numberOfCar: Int, 
    trialCount: Int, 
    strategy: MoveStrategy
) : this( 
    List(numberOfCar) { 
        Driver(Car(), strategy) 
    }, 
    trialCount 
)
```

> secondary constructor를 사용하면 테스트할 때 의도한 데이터만을 가지고 테스트함을 확실히 보장할 수 있다는 장점이 있습니다.

추가로, https://github.com/next-step/kotlin-racingcar/pull/539#discussion_r873731930 에서 고민하시던 내용과도 연관이 있을 것 같아요.

wisemuji commented on src/main/kotlin/racingcar/car/strategy/NameLengthLimitStrategy.kt:
```diff
@@ -0,0 +1,19 @@
+
+import racingcar.car.exception.InvalidDriverNameException
+
+class NameLengthLimitStrategy : NameStrategy {
```
전략 패턴 활용 좋습니다! 👍

wisemuji commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,16 @@
+
+import racingcar.utils.InputParser
+
+object InputView {
+    private const val CAR_NAME_DILIMITER = ","
+    fun getNumberOfCarByConsole(): List<String> {
+        println("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).\n")
+        return readln().split(CAR_NAME_DILIMITER)
+    }
+
+    fun getTrialCountByConsole(): Int {
+        println("시도할 횟수는 몇 회 인가요?")
+        return InputParser.toInt(readln())
+    }
+}
```
https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r871409047 코멘트 확인 부탁드립니다!

wisemuji reviewed (changes requested):
Step4 고생 많으셨습니다! 👏  
꾸준하게 열심히 PR 올려주셔서 넘 보기 좋습니다! 이제 미션 완료가 코앞이네요 ㅎㅎ

각각의 객체가 역할과 책임을 어떻게 가지는게 좋을지 고민하고 잘 구현해주신게 보입니다 ㅎㅎ
고민해보시면 좋을 코멘트들을 남겼습니다.

Step3에서 제가 코멘트를 많이 남겨드려서 그런지 일부 반영되지 않은 코멘트들이 보이네요ㅠㅠ 확인 부탁드립니다!

wisemuji commented:
> tdd 하는 것 처럼
선 테스트 작성 -> 후 로직 구현 & 검증
행동하기가 쉽지 않네요..
쉽지 않은 이유중 하나가.. 커밋단위 컴파일 에러가 많이 발생하는것도 이유인것 같아요.!
커밋 단위 실행하면, 실행 안되는 케이스가 있을까 걱정이 되는데, 이점 참고 부탁드려요.!

TDD 하는 것도 쉽지 않고 그걸 커밋으로 드러내기도 어렵죠ㅠㅠ 
앞으로 교육 진행하시다보면 라이브 강의에서 재성님이 TDD 라이브 코딩을 시연해주실 때가 있을텐데 그게 도움이 많이 되실거라고 생각해요! (아마 로또부터 해주실 것 같아요)
지금도 너무 잘 해주고 계십니다 ☺️

kwanguuuu committed:
[step4] fix: 함수명 변경 (getMoveResult -> getMoveResults )

kwanguuuu committed:
[step4] fix: DriversTest.kt 정상 동작하도록 수정

- 리뷰 반영:https://github.com/next-step/kotlin-racingcar/pull/539#discussion_r873098044

kwanguuuu committed:
[step4] refactor: assertThatThrownBy 를 assertThrows로 변경( NumberInputTest.kt)

- 리뷰 반영: https://github.com/next-step/kotlin-racingcar/pull/539#discussion_r873098287

kwanguuuu committed:
[step4] refactor: isPlayable 이 boolean 값만 반환하도록 수정

- 리뷰 반영: https://github.com/next-step/kotlin-racingcar/pull/539#discussion_r873098226

kwanguuuu commented:
지난번 놓친 리뷰들도 챙겨주셔서 감사합니다.ㅎㅎ
이번 코드 수정하면서 좋은 생각 많이했네요 감사합니다.

코멘트 남긴거 확인 부탁드릴게요! 🙏🏻

wisemuji commented on src/test/kotlin/racingcar/car/AlwaysTrueMoveStrategy.kt:
```diff
@@ -1,6 +1,6 @@
-import step3.racingcar.car.strategy.MoveStrategy
+import racingcar.car.strategy.MoveStrategy
 
 class AlwaysTrueMoveStrategy : MoveStrategy {
```
AlwaysTrueMoveStrategy처럼 전략 클래스를 상속받아 만든 로직과 그렇지 않은 로직은 어떤 차이가 있을까요?

wisemuji commented on src/test/kotlin/racingcar/car/RacingGameTest.kt:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+
+    @Test
+    fun `게임이 진행 가능 횟수가 0 초과할 때 게임을 할 수 있다`() {
+        val correctRacingGameInit = RacingGame(listOf("car1", "car2"), 1, moveStrategy, nameStrategy)
+        val badRacingGameInitCase = RacingGame(listOf("car1", "car2"), 0, moveStrategy, nameStrategy)
+        val badRacingGameInitCase2 = RacingGame(listOf("car1", "car2"), -1, moveStrategy, nameStrategy)
+
+        assertThat(correctRacingGameInit.isPlayable()).isTrue
+        assertThat(badRacingGameInitCase.isPlayable()).isFalse
+        assertThat(badRacingGameInitCase2.isPlayable()).isFalse
+    }
```
`@ParameterizedTest`를 활용해보시면 어떨까요?

wisemuji commented on src/test/kotlin/racingcar/car/MoveResultTest.kt:
```diff
@@ -0,0 +1,35 @@
+class MoveResultTest {
+
+    @Test
+    fun `가장 먼 거리의 차 이름을 반환한다`() {
```
어떤 로직을 테스트하고 싶은지가 누락된 것 같습니다.

wisemuji reviewed (approved):
4단계 넘 고생 많으셨습니다!! 리뷰를 잘 반영해주셨어요 ㅎㅎ 
처음에 비해 많이 성장하신게 보입니다.

특히 제가 드린 코멘트를 바로 반영하는 것이 아닌 충분히 고민해보고 제 의견도 물어보시면서 더 학습하려고 하시는 모습이 인상적입니다 ☺️ 
광유님이 물어보신 질문들에 대한 답변과 고민해보시면 좋을 내용을 코멘트로 남겼으니 충분히 고민해보시고 다음 단계에서 반영해주세요 🙂 (광유님이 다음 단계에서 진행하고 싶다고 하신 내용들도요!)

빠른 미션 진행을 위해 이 PR은 머지하겠습니다 👍
